lib/SPIRV/SPIRVTypeScavenger.cpp

```diff
@@ -0,0 +1,561 @@
+// IR for the output SPIR-V file after LLVM IR completes its transition to
+// opaque pointers.
+//
+//===----------------------------------------------------------------------===//
+
+#include "SPIRVTypeScavenger.h"
+#include "SPIRVInternal.h"
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/NoFolder.h"
+#include "llvm/IR/Operator.h"
+
+#define DEBUG_TYPE "type-scavenger"
+
+using namespace llvm;
+
+static Type *getPointerUseType(Function *F, Op Opcode, unsigned ArgNo);
+void SPIRVTypeScavenger::typeModule(Module &M) {
+  // If typed pointers are in effect, we need to do nothing here.
+  if (M.getContext().supportsTypedPointers())
+    return;
+
+  // Try to fill in any known types for function parameters.
+  for (auto &F : M.functions()) {
+    deduceDeclarationTypes(F);
+  }
+
+  // Collect types for all pertinent values in the module.
+  for (auto &F : M.functions()) {
+    for (Argument &Arg : F.args())
+      if (Arg.getType()->isPointerTy())
+        getTypeInternal(&Arg);
+    for (BasicBlock &BB : F) {
+      for (Instruction &I : BB) {
+        if (I.getType()->isPointerTy())
+          getTypeInternal(&I);
+        correctUseTypes(I);
+      }
+    }
+  }
+
+  // Go through all of the types we have collected, and if any are still
+  // deferred, assign them a fallback i8* type.
+  Type *Int8Ty = Type::getInt8Ty(M.getContext());
+  for (const auto &Pair : DeducedTypes) {
+    if (auto *Untyped = dyn_cast<DeferredType *>(Pair.second)) {
+      LLVM_DEBUG(dbgs() << "No inferrable type for " << *Pair.first << "\n");
+      fixType(*Untyped, Int8Ty);
+      DeducedTypes[Pair.first] = Int8Ty;
+    }
+  }
+  return;
+}
+
+static Type *getPointerUseType(Function *F, Op Opcode, unsigned ArgNo) {
+  switch (Opcode) {
+  case OpAtomicLoad:
+  case OpAtomicExchange:
+  case OpAtomicCompareExchange:
+  case OpAtomicIAdd:
+  case OpAtomicISub:
+  case OpAtomicFAddEXT:
+  case OpAtomicSMin:
+  case OpAtomicUMin:
+  case OpAtomicFMinEXT:
+  case OpAtomicSMax:
+  case OpAtomicUMax:
+  case OpAtomicFMaxEXT:
+  case OpAtomicAnd:
+  case OpAtomicOr:
+  case OpAtomicXor:
+    if (ArgNo == 0)
+      return F->getReturnType();
+    return nullptr;
+  case OpAtomicStore:
+    if (ArgNo == 0)
+      return F->getArg(3)->getType();
+    return nullptr;
+  default:
+    return nullptr;
+  }
+}
+
+void SPIRVTypeScavenger::typeIntrinsicArgs(Function &F, Intrinsic::ID Id) {
+  static constexpr unsigned Return = ~0U;
+  auto AddParameter = [&](unsigned ArgNo, DeducedType Ty) {
+    if (ArgNo == Return) {
+    } else {
+      Argument *Arg = F.getArg(ArgNo);
+      LLVM_DEBUG(dbgs() << "Parameter " << *Arg << " of " << F.getName() << " has type " << Ty << "\n");
+      DeducedTypes[Arg] = Ty;
+    }
+  };
+  LLVMContext &Ctx = F.getContext();
+
+  switch (Id) {
+  // TODO: handle case Intrinsic::memset
```

Comment:
If I understand correctly, at this point we no longer have memsets in IR since they were transformed to memcpys

Reply:
I was using the list of intrinsics in SPIRVWriter's `transIntrinsicInst` as the base, which does include `memset` as a handled intrinsic.

Reply:
Yeah, I mean there is no need to handle memset, since the Regularize LLVM pass has already replaced memsets with memcpys and hence the comment can be removed :)